rfcs/45-stageless.md

```diff
@@ -0,0 +1,866 @@
+   // for all systems in the schedule.
+   // All labels are converted to specific systems.
+   // `Graph` is a bikeshed-avoidance graph storage structure,
+   // the exact strategy should be benchmarked
+   ordering_constraints: Graph<SystemId>,
+   // Label configuration must be stored at the schedule level,
+   // rather than attached to specific label instances
+   labels: HashMap<Box<dyn SystemLabel>, LabelConfig>,
+   // Used to check if mutation is allowed
+   currently_running: bool,
+}
+```
+
+Configured systems are composed of a raw system and its configuration.
+
+```rust
+struct ConfiguredSystem{
+   raw_system: System,
+   config: SystemConfig,
+}
+```
+
+Raw systems store metadata, cached state, `Local` system parameters and the function pointer to the actual function to be executed:
+
+```rust
+struct System {
+   function: F,
+   // Contains all metadata and state
+   meta: SystemMeta
+   last_change_tick: u32,
+}
+```
+
+System configuration stores ordering dependencies, run criteria and labels.
+It can be created raw, or stored in association with either system labels or systems.
+
+```rust
+enum OrderingTarget {
+   Label(Box<dyn SystemLabel>),
+   System(SystemId),
+}
+
+struct SystemConfig {
+   strict_ordering_before: Vec<OrderingTarget>,
+   strict_ordering_after: Vec<OrderingTarget>,
+   // The `TypeId` here is the type id of the flushing system
+   flushed_ordering_before: Vec<(TypeId, OrderingTarget)>,
+   flushed_ordering_after: Vec<(TypeId, OrderingTarget)>,
+   run_criteria: Vec<SystemId>,
+   // The combined access requirements of the system and its run criteria
+   joint_access: FilteredAccessSet<ComponentId>,
+   // This vector is always empty when this struct is used for labels
+   labels: Vec<Box dyn SystemLabel>,
+}
+
+struct LabelConfig {
+   system_config: SystemConfig,
+   allow_ambiguities: bool,
+}
+
+```
+
+### Label configuration
+
+When a property is applied to a label, it is ultimately applied to each system in the label, as if it had been labelled individually.
+Of course, properties that only make sense at a group level (such as whether ambiguities within that label are allowed) are not propagated down, as they are not defined for inidividual systems.
+
+When building the constraint graph, all labels are expanded into the set of systems that they contain.
+Thus, adding an ordering constraint between label N and M is the same as adding a constraint between each `n * m` pair of systems.
+
+### Run criteria
+
+Run criteria can access arbitrary data from the world, have their own fetches and are run on the world.
+While they reuse much of the system machinery, they have several important constraints:
+
+- they cannot mutate data
+  - required to avoid having to order run criteria relative to each other
+- they cannot capture or store local data
+  - required to avoid user footguns
+  - required to efficiently special-case their execution

```

Comment:
I don't think it's as serious of a footgun, and opens up some interesting patterns, like "run every other time". It's also relatively tricky to enforce. IMO we should permit it.

Reply:
I did more thinking and came to the conclusion that it's still better to not allow them. There is footgun potential because things can have multiple RC attached.

If the "run every other time" RC used a `Local`, it would work if it was alone. But if it was paired with another RC, the local would eagerly update even if the other RC fails and the system doesn't run.

I think a totally-safe "run every other time" would follow the cached resource pattern. We'd have an RC check some counter resource and only return `true` if `count % 2 == 0`.